src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php

```diff
@@ -0,0 +1,376 @@
+        }
+
+        if ($choices instanceof \Traversable) {
+            $choices = iterator_to_array($choices);
+        }
+
+        // If the choices are given as recursive array (i.e. with explicit
+        // choice groups), flatten the array. The grouping information is needed
+        // in the view only.
+        self::flatten($choices, $flatChoices);
+
+        return new ArrayChoiceList($flatChoices, $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @deprecated Added for backwards compatibility in Symfony 2.7, to be
+     *             removed in Symfony 3.0.
+     */
+    public function createListFromFlippedChoices($choices, $value = null)
+    {
+        if (!is_array($choices) && !$choices instanceof \Traversable) {
+            throw new UnexpectedTypeException($choices, 'array or \Traversable');
+        }
+
+        if ($choices instanceof \Traversable) {
+            $choices = iterator_to_array($choices);
+        }
+
+        // If the choices are given as recursive array (i.e. with explicit
+        // choice groups), flatten the array. The grouping information is needed
+        // in the view only.
+        self::flattenFlipped($choices, $flatChoices);
+
+        // If no values are given, use the choices as values
+        // Since the choices are stored in the collection keys, i.e. they are
+        // strings or integers, we are guaranteed to be able to convert them
+        // to strings
+        if (null === $value) {
+            $value = function ($choice) {
+                return (string) $choice;
+            };
+        }
+
+        return new ArrayKeyChoiceList($flatChoices, $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createListFromLoader(ChoiceLoaderInterface $loader, $value = null)
+    {
+        if (null !== $value && !is_callable($value)) {
+            throw new UnexpectedTypeException($value, 'null or callable');
+        }
+
+        return new LazyChoiceList($loader, $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null)
+    {
+        if (null !== $preferredChoices && !is_array($preferredChoices) && !is_callable($preferredChoices)) {
+            throw new UnexpectedTypeException($preferredChoices, 'null, array or callable');
+        }
+
+        if (null !== $label && !is_callable($label)) {
+            throw new UnexpectedTypeException($label, 'null or callable');
+        }
+
+        if (null !== $index && !is_callable($index)) {
+            throw new UnexpectedTypeException($index, 'null or callable');
+        }
+
+        if (null !== $groupBy && !is_array($groupBy) && !$groupBy instanceof \Traversable && !is_callable($groupBy)) {
+            throw new UnexpectedTypeException($groupBy, 'null, array, \Traversable or callable');
```

Comment:
Since the types are checked in ChoiceType, it should be fine to drop these checks